### PR TITLE
(rfc) chat models

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: langchain/chat_models/base.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class SimpleChatModel(BaseChatModel):
+    role: str = "assistant"
+
+    def _generate(
+        self, messages: List[BaseMessage], stop: Optional[List[str]] = None
+    ) -> ChatResult:
+        output_str = self._call(messages, stop=stop)
+        message = BaseMessage(text=output_str, role=self.role)
```
Comment (agola11) [inline langchain/chat_models/base.py:40]: Why don't we just get rid of role and use AIMessage

File: langchain/chat_models/openai.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    environment variable ``OPENAI_API_KEY`` set with your API key.
+
+    Any parameters that are valid to be passed to the openai.create call can be passed
+    in, even if not explicitly saved on this class.
+
+    Example:
+        .. code-block:: python
+
+            from langchain.chat import OpenAI
+            openai = OpenAI(model_name="gpt-3.5-turbo")
```
Comment (agola11) [inline langchain/chat_models/openai.py:42]: ChatOpenAI

File: langchain/chat_models/openai.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        params: Dict[str, Any] = {**{"model": self.model_name}, **self._default_params}
+        if stop is not None:
+            if "stop" in params:
+                raise ValueError("`stop` found in both the input and default params.")
+            params["stop"] = stop
+        message_dicts = [self._convert_message_to_dict(m) for m in messages]
+        response = self.completion_with_retry(messages=message_dicts, **params)
+        generations = []
+        for res in response["choices"]:
+            message = self._convert_message_to_dict(res["message"])
```
Comment (agola11) [inline langchain/chat_models/openai.py:169]: I think you meant _convert_dict_to_message here

File: langchain/prompts/chat.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+from langchain.schema import (
+    AIMessage,
+    BaseMessage,
+    ChatMessage,
+    HumanMessage,
+    SystemMessage,
+)
+
+
+class BaseMessagePromptTemplate(BaseModel, ABC):
```
Comment (agola11) [inline langchain/prompts/chat.py:20]: How about shortening these class names to be `*PromptMessage`

File: langchain/prompts/chat.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            ChatMessagePromptTemplate(
+                text=PromptTemplate.from_template(template), role=role
+            )
+            for role, template in string_messages
+        ]
+        input_vars = set([m.prompt.input_variables] for m in messages)
+        return cls(input_variables=list(input_vars), messages=messages)
+
+    @classmethod
+    def from_strings(
```
Comment (agola11) [inline langchain/prompts/chat.py:84]: I would argue that a fromBaseMessagePromptTemplate (i.e.
```
@classmethod
def fromPromptMessages(cls, List[BaseMessagePromptTemplate]):
  ...
```
is the most useful. If we want to use these other two classmethods, they can just call it after constructing the PromptMessages

File: langchain/chat_models/base.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+class BaseChatModel(ABC):
+    def generate(
+        self, messages: List[List[BaseMessage]], stop: Optional[List[str]] = None
+    ) -> LLMResult:
+        """Top Level call"""
+        results = []
+        for m in messages:
+            results.append(self._generate(m, stop=stop))
+        return LLMResult(generations=[res.generations for res in results])
```
Comment (agola11) [inline langchain/chat_models/base.py:15]: Am I missing something here? I don't think this should work. `LLMResult` expects a `List[List[Generation]]`
and `ChatResult` has a `List[ChatGeneration]`, we need to either allow both `ChatGeneration` or `Generation` in `LLMResult` or do something else

File: langchain/chat_models/openai.py [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    in, even if not explicitly saved on this class.
+
+    Example:
+        .. code-block:: python
+
+            from langchain.chat_models import ChatOpenAI
+            openai = ChatOpenAI(model_name="gpt-3.5-turbo")
+    """
+
+    client: Any  #: :meta private:
```
Comment (colinalexander) [inline langchain/chat_models/openai.py:110]: My linter is complaining that I am not specifying a value for `client` when instantiating this class.  I assume the following would be OK?
```
client: Any = None

**Pull request merged**